hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/CopyFromLocalOperation.java

```diff
@@ -0,0 +1,241 @@
+    private final boolean overwrite;
+    private final Path source;
+    private final Path destination;
+
+    private FileStatus dstStatus;
+
+    public CopyFromLocalOperation(
+            final StoreContext storeContext,
+            Path source,
+            Path destination,
+            boolean deleteSource,
+            boolean overwrite,
+            CopyFromLocalOperationCallbacks callbacks) {
+        super(storeContext);
+        this.callbacks = callbacks;
+        this.deleteSource = deleteSource;
+        this.overwrite = overwrite;
+        this.source = source;
+        this.destination = destination;
+    }
+
+    @Override
+    @Retries.RetryTranslated
+    public Void execute()
+            throws IOException, PathExistsException {
+        LOG.debug("Copying local file from {} to {}", source, destination);
+        File sourceFile = callbacks.pathToFile(source);
+        try {
+            dstStatus = callbacks.getFileStatus(destination);
+        } catch (FileNotFoundException e) {
+            dstStatus = null;
+        }
+
+        checkSource(sourceFile);
+        prepareDestination(destination, sourceFile, overwrite);
+        uploadSourceFromFS();
+
+        if (deleteSource) {
+            callbacks.delete(source, true);
+        }
+
+        return null;
+    }
+
+    private void uploadSourceFromFS()
+            throws IOException, PathExistsException {
+        RemoteIterator<LocatedFileStatus> localFiles = callbacks
+                .listStatusIterator(source, true);
+
+        // After all files are traversed, this set will contain only emptyDirs
+        Set<Path> emptyDirs = new HashSet<>();
+        List<UploadEntry> entries = new ArrayList<>();
+        while (localFiles.hasNext()) {
+            LocatedFileStatus sourceFile = localFiles.next();
+            Path sourceFilePath = sourceFile.getPath();
+
+            // Directory containing this file / directory isn't empty
+            emptyDirs.remove(sourceFilePath.getParent());
+
+            if (sourceFile.isDirectory()) {
+                emptyDirs.add(sourceFilePath);
+                continue;
+            }
+
+            Path destPath = getFinalPath(sourceFilePath);
+            // UploadEntries: have a destination path, a file size
+            entries.add(new UploadEntry(
+                    sourceFilePath,
+                    destPath,
+                    sourceFile.getLen()));
+        }
+
+        if (localFiles instanceof Closeable) {
+            ((Closeable) localFiles).close();
+        }
+
+        // Sort all upload entries based on size
+        entries.sort(new ReverseComparator(new UploadEntry.SizeComparator()));
+
+        int LARGEST_N_FILES = 5;
+        final int sortedUploadsCount = Math.min(LARGEST_N_FILES, entries.size());
+        List<UploadEntry> uploaded = new ArrayList<>();
+
+        // Take only top most X entries and upload
+        for (int uploadNo = 0; uploadNo < sortedUploadsCount; uploadNo++) {
```

Comment:
Is this parallelized? I know that the xfer manager does it across threads, but this is still uploading one by one. This is probably simplest to start with...

Reply:
Not parallelized yet, wasn't clear on the best way to do it in the code base so left it simple for the first pass. The reason why I wasn't clear is, as you said xfer manager does it across threads depending on file size, which capacity the executor should work with. 

Is there any other executor other than the one available from `ContextStore` I should be aware of?  And do you have any thoughts around its capacity?

Reply:
Using an executor for them now, let me know what you think.

Reply:
If we use the xfer manager -which is the simplest- it gets one of the thread pools we create. It has limitations, but we haven't sat down to replace it (yet) as it mostly works OK. 


For other uses, the StoreContext executors should be the executor pool. And it's possible to create a restricted subpool of that, which is what S3ABlockOutputStream does.

FWIW, I am not sure that (a) the default size of that pool is right and (b) whether we should just go for for an unlimited pool. Big changes, but as most of the threads are blocked on remote IO, pool size is probably a needless limitation.